File: clients/kaiju/network/models/protocol.go
package models

//fomoxa:model codec=game
type ClientHello struct {
	ClientKind  uint8  `fomoxa:"u8" codec:"game"`
	DisplayName string `fomoxa:"string" codec:"game"`
}

//fomoxa:model codec=game
type Welcome struct {
	PlayerID      uint32  `fomoxa:"u32" codec:"game"`
	PlaneHalfSize float32 `fomoxa:"f32" codec:"game"`
	TickRate      uint16  `fomoxa:"u16" codec:"game"`
}

//fomoxa:model codec=game
type PlayerInput struct {
	Sequence  uint32  `fomoxa:"u32" codec:"game"`
	MoveX     float32 `fomoxa:"f32" codec:"game"`
	MoveZ     float32 `fomoxa:"f32" codec:"game"`
	Jump      bool    `fomoxa:"bool" codec:"game"`
	LookYaw   float32 `fomoxa:"f32" codec:"game"`
	LookPitch float32 `fomoxa:"f32" codec:"game"`
}

//fomoxa:model codec=game
type PlayerState struct {
	PlayerID   uint32  `fomoxa:"u32" codec:"game"`
	ClientKind uint8   `fomoxa:"u8" codec:"game"`
	Color      uint32  `fomoxa:"u32" codec:"game"`
	PositionX  float32 `fomoxa:"f32" codec:"game"`
	PositionZ  float32 `fomoxa:"f32" codec:"game"`
	PositionY  float32 `fomoxa:"f32" codec:"game"`
	LookYaw    float32 `fomoxa:"f32" codec:"game"`
	LookPitch  float32 `fomoxa:"f32" codec:"game"`
}

//fomoxa:model codec=game
type WorldSnapshot struct {
	Tick    uint32        `fomoxa:"u32" codec:"game"`
	Players []PlayerState `fomoxa:"Array<PlayerState>" codec:"game"`
}

File: server/src/models/protocol.rs
use fomoxa_attributes::*;

#[network]
#[codec(game)]
#[derive(Debug, Default, Clone, PartialEq)]
pub struct ClientHello {
    #[network(u8)]
    #[codec(game)]
    pub client_kind: u8,

    #[network(string)]
    #[codec(game)]
    pub display_name: String,
}

#[network]
#[codec(game)]
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Welcome {
    #[network(u32)]
    #[codec(game)]
    pub player_id: u32,

    #[network(f32)]
    #[codec(game)]
    pub plane_half_size: f32,

    #[network(u16)]
    #[codec(game)]
    pub tick_rate: u16,
}

#[network]
#[codec(game)]
#[derive(Debug, Default, Clone, PartialEq)]
pub struct PlayerInput {
    #[network(u32)]
    #[codec(game)]
    pub sequence: u32,

    #[network(f32)]
    #[codec(game)]
    pub move_x: f32,

    #[network(f32)]
    #[codec(game)]
    pub move_z: f32,

    #[network(bool)]
    #[codec(game)]
    pub jump: bool,

    #[network(f32)]
    #[codec(game)]
    pub look_yaw: f32,

    #[network(f32)]
    #[codec(game)]
    pub look_pitch: f32,
}

#[network]
#[codec(game)]
#[derive(Debug, Default, Clone, PartialEq)]
pub struct PlayerState {
    #[network(u32)]
    #[codec(game)]
    pub player_id: u32,

    #[network(u8)]
    #[codec(game)]
    pub client_kind: u8,

    #[network(u32)]
    #[codec(game)]
    pub color: u32,

    #[network(f32)]
    #[codec(game)]
    pub position_x: f32,

    #[network(f32)]
    #[codec(game)]
    pub position_z: f32,

    #[network(f32)]
    #[codec(game)]
    pub position_y: f32,

    #[network(f32)]
    #[codec(game)]
    pub look_yaw: f32,

    #[network(f32)]
    #[codec(game)]
    pub look_pitch: f32,
}

#[network]
#[codec(game)]
#[derive(Debug, Default, Clone, PartialEq)]
pub struct WorldSnapshot {
    #[network(u32)]
    #[codec(game)]
    pub tick: u32,

    #[network(Array<PlayerState>)]
    #[codec(game)]
    pub players: Vec<PlayerState>,
}
